Add timestamps to notifications

cfsmp3 opened the pull request:
I find it a bit annoying that when I open the notification app I just get a list of notifications but no clue about when each arrived so this small PR solves that.

Note - timestamp format is currently hardcoded. We probably want to make it configurable, but it doesn't seem like there's a config file that reaches this code so unsure about the best way to go about it.

cfsmp3 committed:
Add timestamp to notifications

kgilmer commented:
Hi @cfsmp3 , thanks for your contribution!  I agree that sorting by time is the most sane way of rendering a list of notifications.  I notice that you've updated the rofi UI.  Can you share a screenshot of how the additional field looks?

cfsmp3 commented:
> Can you share a screenshot of how the additional field looks?

This is beyond pathetic - I can't take a screenshot when notification window is open :-) So pardon me for just sending a mobile screenshot to avoid spending time now in figuring this out.
![notifications_regolith](https://user-images.githubusercontent.com/5949913/138936058-ac6b1ea0-5d77-405d-85d3-6ccfa8c0e891.jpg)

cfsmp3 committed:
Make datetime format configurable via arg

cfsmp3 commented:
@farsil @kgilmer I made the format configurable. It defaults to locale.

cfsmp3 committed:
Add daemon

kgilmer commented on rofication/_gui.py:
```diff
@@ -36,7 +36,8 @@ def rofi_entry(notification: Notification) -> str:
     stripped_app = strip_tags(notification.application)
     stripped_body = strip_tags(' '.join(notification.body.split()))
-    return f'<b>{stripped_summ}</b> <small>({stripped_app})</small>\n<small>{stripped_body}</small>'
+    stripped_timestamp = f"{strip_tags (notification.timestamp)} " if notification.timestamp else ""
```
I would like this UI addition to be opt-in and user-configurable.  It could be enabled via a commandline parameter or an Xresource key/value pair.

cfsmp3 replied:
Actually it is user configurable. That's what the parameter --tsformat does (on rotification daemon, which is called from regolith/i3/config). 

The default format with this PR is the system locale, but the right place to specify a default and document this parameter would be the config file itself, which for me looks like this:

```
  ###############################################################################
  # Rofication date & time format in Python format, see
  # https://docs.python.org/3/library/datetime.html
  # Defaults to locale default. Use empty string ("") for no timestamp
  ###############################################################################
  set_from_resource $rofication_datetime_format rofication_datetime_format "%x %X"
```
....
```
# Start Rofication for notifications
  set_from_resource $i3-wm.program.notifications i3-wm.program.notifications /usr/bin/rofication-daemon --tsformat $rofication  _datetime_format
  exec --no-startup-id $i3-wm.program.notifications...
```

cfsmp3 replied:
See this sibling PR:
https://github.com/regolith-linux/regolith-i3-gaps-config/pull/38/files

kgilmer replied:
Yes you're right, great.  

I think we can avoid a change in i3-gaps-config (the file you edited will not be shipped in Regolith 2.0 FYI) by directly reading from Xresources in rofication-daemon, and avoiding the command line arg as well.  This will localize the change entirely to rofication-dameon.  See [this utility class](https://github.com/regolith-linux/regolith-rofication/blob/master/rofication/_util.py#L20) for reading from env and xresources.

kgilmer replied:
FYI this is where i3 config lives in Regolith 2.0: https://github.com/regolith-linux/regolith-i3-config/tree/master/partials

cfsmp3 replied:
Changed.

kgilmer reviewed (changes requested):
By default I want the UI to remain the same.  The new functionality should be enabled if the user specifies so.

cfsmp3 committed:
Read timestamp format from XResources

kgilmer commented on rofication-daemon:
```diff
@@ -1,10 +1,14 @@
 #!/usr/bin/env python3
 
+import argparse
```
seems like this is no longer needed

kgilmer reviewed (approved):
nit on the import, LGTM

cfsmp3 committed:
Remove unneeded import

cfsmp3 committed:
Fix conflicts

cfsmp3 committed:
Merge branch 'master' of https://github.com/regolith-linux/regolith-rofication into regolith-linux-masterS

cfsmp3 committed:
Merge branch 'regolith-linux-master'

kgilmer commented:
It's fine to merge as is, and I pulled down your branch to do a quick test with a fake notification:

```
$ nc -U /tmp/rofi_notification_daemon
list
[{"id": 1, "summary": "Hello! January", "body": "", "application": "notify-send", "icon": "face-wink", "urgency": 1, "actions": [], "hints": {"urgency": 1}, "timestamp": ""}]
```

I'm building a new front-end to rofication, [here](https://github.com/regolith-linux/ilia/tree/main/src/notifications).  I would like default to sort by notification creation time.  It seems to me that the time format should impact the UI, but the internal date format should be something stable and simple (for conversion) so that it can be used generally (such as for sort).  What do you think?  Perhaps the timestamp format should only be read/used on UIs and the daemon should simply represent it as a unit timestamp..

cfsmp3 committed:
Made internal format a Python timestamp, convert to string in the UI

cfsmp3 commented:
> I'm building a new front-end to rofication, [here](https://github.com/regolith-linux/ilia/tree/main/src/notifications). I would like default to sort by notification creation time. It seems to me that the time format should impact the UI, but the internal date format should be something stable and simple (for conversion) so that it can be used generally (such as for sort). What do you think? 

Yes, actually I implemented it this way on the assumption that the format would come from the config file and be passed down to the daemon, the Dbus, etc... 

Reading the format as a Resource simplifies that.

I just updated the PR. The timestamp is now a sortable number (note: a float, since we're in Python) and the GUI makes the actual conversion to string.

About sorting, my plan was to make it configurable as well and just insert at the beginning or the end of the list rather than do an actual sort.

kgilmer reviewed (approved):
Great!  Feel free to merge when you're happy with it.  From there I'll push a bump in the changelog and that will cause the package builder to make it available for testing here: https://github.com/regolith-linux/package-repo

cfsmp3 committed:
Undo unneeded changes in daemon

kgilmer commented:
Nice work @cfsmp3 .  FYI your feature is building into the 2.0 package repository here: https://github.com/regolith-linux/package-repo/actions/runs/1442476657
